src/core/search/search.cc

```diff
@@ -25,65 +27,142 @@ AstExpr ParseQuery(std::string_view query) {
   return driver.Take();
 }
 
-// Add results from matched to current, keep sorted and remove duplicates.
-void UnifyResults(vector<DocId>&& matched, vector<DocId>* current, vector<DocId>* tmp) {
-  swap(*current, *tmp);  // Move current to tmp so we can directly write into current
-  current->clear();
-  current->reserve(matched.size() + tmp->size());
+// Represents an either owned or non-owned result set that can be accessed transparently.
+struct IndexResult {
+  using DocVec = vector<DocId>;
+
+  IndexResult() : value{DocVec{}} {};
+
+  IndexResult(const DocVec* dv) : value{dv} {
+    if (dv == nullptr)
+      value = DocVec{};
+  }
+
+  IndexResult(DocVec&& dv) : value{move(dv)} {
+  }
+
+  // Transparent const access to underlying value
+  const DocVec* operator->() const {
+    return holds_alternative<DocVec>(value) ? &get<DocVec>(value) : get<const DocVec*>(value);
+  }
+
+  // Move out of owned or copy borrowed
+  DocVec Take() {
+    if (holds_alternative<DocVec>(value))
+      return move(get<DocVec>(value));
+    return *get<const DocVec*>(value);
+  }
+
+  // Move out of owned or return empty
+  DocVec TakeOwnedOrEmpty() {
+    if (holds_alternative<DocVec>(value))
+      return move(get<DocVec>(value));
+    return DocVec{};
+  }
+
+  static bool BySize(const IndexResult& l, const IndexResult& r) {
+    return l->size() < r->size();
+  }
+
+ private:
+  variant<DocVec /*owned*/, const DocVec* /*pointer to borrowed*/> value;
```

Comment:
An alternative (which I think would be more readable) is to have an owning member of type `DocVec`, and then another member of type `const DocVec*` that either points to it, or to some other thing.
Your call if you'd like to go that path

Reply:
What about moving 🚗 , copying, etc ... I'd have to fix it for the self-referencing pointer.  Variant is an elegant way to show that only one of both can be stored and that manages the value automatically

Having two members of which only one is active and the other is a pointer that references another member reminds me of old-style messy c++

Reply:
I'm all for using `variant` as well, but here I did not suggest that only one will be active. I meant that the pointer of them will always be active, and the non-pointer will only passively hold data if needed.
I guess it's somewhat similar to our `GetSlice()` with the `tmp` var just holding data when needed

Reply:
But, as always, if you don't like this idea, just ignore it :)